### ISSUE TITLE
Lighthouse not producing city connection to capital

### ISSUE THREAD

**Issue by austin-wangyu** [LoneGazebo/Community-Patch-DLL#4427]
_1. Mod version (i.e Date - 4/23):_
Most current as of 5/23/18

_2. Mod list (if using Vox Populi only, leave blank):_

_3. Error description:_ 
See title, though each city has lighthouse, produces unhappiness due to lack of city connection as well as lack of income from city connection. Fish and other resources receive the lighthouse food and gold bonus, though.

_4. Steps to reproduce (optional):_
As Carthage, settle a whole bunch of cities along the coast. No clue why they don't work, they can't all be blockaded at the same time this early on.

---------------------------
Supporting information:
Please note that you can attach .zip files by dragging-and-dropping them. If possible, zip up all supporting data and post that way.

1. Log files (always attach your Logs folder, located at My Documents/My Games/Sid Meier's Civilization 5. Make sure you have enabled logging before experiencing an error! Go here to find out how: http://forums.civfanatics.com/showthread.php?t=487482):
Unfortunately did not have logging enabled before this. Have enabled now for future reference.

2. Save game (always attach a save that was made a turn before the error; located at My Documents/My Games/Sid Meier's Civilization 5/ModdedSaves):
[Before lighthouse glitch.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/2030003/Before.lighthouse.glitch.zip)

3. CvMiniDump.dmp file (attach if experiencing a game crash. Located at Program Files/Steam/steamapps/common/Sid Meier's Civilization V):

4. Screenshots (optional):
![image](https://user-images.githubusercontent.com/22928672/40410847-990eb406-5e23-11e8-8605-a5457460af60.png)

**Comment by austin-wangyu** [comment]
Update: even after unlocking Lighthouse tech it still didn't resolve.

**Comment by Mzimoxs** [comment]
In Vanilla Civ5 you need the road tech to unlock city connections (so wheel)

**Comment by ilteroi** [comment]
can't reproduce.

![image](https://user-images.githubusercontent.com/1358966/40444307-4ebb2ba4-5ec9-11e8-91c2-3a79ebbd940b.png)

there's basically only two conditions: you need a path of revealed coastal water plots to the capital.

that explains the problem with hippo regius. for utique it's hard to tell, maybe explore a bit more around the peninsula to the south?

**Comment by ryanmusante** [comment]
I play exclusively with Carthage. The route to the south end does not a sea connection to Carthage. There is still fog of war covering the coast tiles that will connect and create a lighthouse upon revealing them.

![_20180524_061640](https://user-images.githubusercontent.com/26808646/40451670-53fc97aa-5f1a-11e8-94f4-016f85829922.JPG)

**Comment by Mzimoxs** [comment]
The coastline to Utique is already revealed. The fog of war is over the tiles under the coast tiles.

**Comment by ryanmusante** [comment]
There is a single tile uncovered at Utique that would allow the sea connection. It is to the lower left as highlighted in green.

**Comment by Mzimoxs** [comment]
The tiles you highlighted ARE uncovered. You wouldn't see the tile yields otherwise.

**Comment by ryanmusante** [comment]
I'm talking about the the tiles next to it that are not showing yields, especially the one center bottom. It is the reason why there is no sea connection to Utique.

**Comment by Iamblichos** [comment]
Here's a test to verify @ilteroi. Open up your save from the screenshot above. Have IGE loaded, place some boats in the water to reveal those tiles, do turn. Problem fixed? Should be.

**Comment by supermerill** [comment]
They seems revealed for utique. Maybe because the CS isn't allied with you and so, it blocks the path?

**Comment by ilteroi** [comment]
enemy territory doesn't block trade routes.

those two tiles are not really revealed yet, it's just a trick of the camera.

**Comment by Mzimoxs** [comment]
Again: They are revealed. Else you wouldn't see the yields (2 food for coast).

**Comment by ryanmusante** [comment]
But does showing yields count technically as being "revealed" to the AI? Or does it require the entire tile to be unmasked from the cloud for the properties to be available?

**Comment by Albertolona** [comment]
Look more carefully at his screenshot and notice the real reason why Utique is not connecting.

Utique is NOT a coastal city so he can't have built a lighthouse there.

**Comment by Mzimoxs** [comment]
yes it is. The tile 7 from Utique is coastal.
And also that wouldn't explain Gades and Saguntum

**Comment by Albertolona** [comment]
Yes I see that now thanks.

**Comment by ilteroi** [comment]
wow quite the thread for a rather simple issue. can anyone load the save, put an explorer on the peninsula and see what happens please?

**Comment by Iamblichos** [comment]
@ilteroi I tried to load his save but was unsuccessful. What version of VP are you using AND are you using any other mods??

**Comment by austin-wangyu** [comment]
4-20-5, no other mods

**Comment by ilteroi** [comment]
well, now i'm confused. so it was alright before the CS border expanded? seems obvious, but here's the code:

```
//	--------------------------------------------------------------------------------
int TradeRouteWaterValid(const CvAStarNode* parent, const CvAStarNode* node, const SPathFinderUserData&, const CvAStar* finder)
{
	if(parent == NULL)
		return TRUE;

	const TradePathCacheData* pCacheData = reinterpret_cast<const TradePathCacheData*>(finder->GetScratchBuffer());

	CvMap& kMap = GC.getMap();
	CvPlot* pNewPlot = kMap.plotUnchecked(node->m_iX, node->m_iY);

	if (!pNewPlot->isRevealed(pCacheData->GetTeam()))
		return FALSE;

	//ice in unowned territory is not allowed
	if (pNewPlot->isIce() && !pNewPlot->isOwned())
		return FALSE;

	//ocean needs trait or tech
	if (pNewPlot->isDeepWater())
		return pCacheData->CanCrossOcean();

	//coast is always ok
	if (pNewPlot->isShallowWater())
		return TRUE;

	//check passable improvements
	if(pNewPlot->isCityOrPassableImprovement(pCacheData->GetPlayer(),false) && pNewPlot->isAdjacentToShallowWater() )
		return TRUE;

	return FALSE;
}
```

no check for tile ownership. i'm stumped.

**Comment by austin-wangyu** [comment]
@LoneGazebo I'm not sure which coast tile you're referring to?

**Comment by ilteroi** [comment]
hey guys, i think i was looking in the wrong place. sorry, the correct piece of code is

```
int WaterRouteValid(const CvAStarNode* parent, const CvAStarNode* node, const SPathFinderUserData& data, const CvAStar*)
{
	if(parent == NULL)
		return TRUE;

	PlayerTypes ePlayer = data.ePlayer;
	TeamTypes eTeam = GET_PLAYER(ePlayer).getTeam();

	CvPlot* pNewPlot = GC.getMap().plotUnchecked(node->m_iX, node->m_iY);

	if(!pNewPlot || !pNewPlot->isRevealed(eTeam))
		return FALSE;

	if(pNewPlot->getOwner()!=NO_PLAYER && !pNewPlot->IsFriendlyTerritory(ePlayer))
		return FALSE;

	CvCity* pCity = pNewPlot->getPlotCity();
	if(pCity && pCity->getTeam() == eTeam)
		return TRUE;

	if(pNewPlot->isWater())
		return TRUE;

	return FALSE;
}
```

so there is an ownership check after all. the question is if minor players' territory should be excluded from the check, ie allowed for routes?

**Comment by bigcat88** [comment]
@ilteroi no, it was always like that. If someone need a road throw CS then ally it or make friendly.

**Comment by LoneGazebo** [comment]
@ilteroi since roads overrule CS ownership for connections, and you can't build roads at sea, I feel like CS tile ownership shouldn't affect water routes.

**Comment by ilteroi** [comment]
uuh i smell a forum poll

**Comment by LoneGazebo** [comment]
let's just change it and change it back if people balk. I feel like this a QoL change.

**Comment by ilteroi** [comment]
i just took a look but it seems roads trump plot ownership only if there's an active quest to build a road.

**Comment by LoneGazebo** [comment]
@ilteroi hmm...well, then maybe we leave it. I thought roads overwrote cs ownership in all instances.

**Comment by Iamblichos** [comment]
I'm with @LoneGazebo change it. But I like changes. Seriously though, seems stupid that CS ownership of a tile would block water if !AtWar()

**Comment by ilteroi** [comment]
well then this isn't only about water connections anymore.

but i agree, the logical thing would be to allow connections through plots you can enter peacefully. so minor and major with open borders.
